ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/KeyValueCommittingSpec.scala

```diff
@@ -0,0 +1,201 @@
+import org.scalatest.matchers.should.Matchers
+import org.scalatest.wordspec.AnyWordSpec
+import java.time.Instant
+
+class KeyValueCommitingSpec extends AnyWordSpec with Matchers {
+  private val metrics: Metrics = new Metrics(new MetricRegistry)
+  private val keyValueSubmission = new KeyValueSubmission(metrics)
+
+  private val alice = Party.assertFromString("Alice")
+  private val commandId = CommandId.assertFromString("cmdid")
+
+  private def toSubmission(tx: SubmittedTransaction) = {
+    val timestamp = Time.Timestamp.assertFromInstant(Instant.EPOCH)
+    val meta = TransactionMeta(
+      ledgerEffectiveTime = timestamp,
+      workflowId = None,
+      submissionTime = timestamp,
+      submissionSeed = crypto.Hash.hashPrivateKey(this.getClass.getName),
+      optUsedPackages = Some(Set.empty),
+      optNodeSeeds = None,
+      optByKeyNodes = None,
+    )
+    val submitterInfo = SubmitterInfo(
+      actAs = List(alice),
+      applicationId = ApplicationId.assertFromString("appid"),
+      commandId = commandId,
+      deduplicateUntil = Instant.EPOCH,
+    )
+    keyValueSubmission.transactionToSubmission(
+      submitterInfo = submitterInfo,
+      meta = meta,
+      tx = tx,
+    )
+  }
+
+  def getOutputs(builder: TransactionBuilder) =
+    KeyValueCommitting.submissionOutputs(toSubmission(builder.buildSubmitted()))
+
+  val dedupKey = DamlStateKey.newBuilder
+    .setCommandDedup(
+      DamlCommandDedupKey.newBuilder.addSubmitters(alice).setCommandId(commandId).build
+    )
+    .build
+
+  private val keyValue = Value.ValueUnit
+  private val templateId = Identifier.assertFromString("pkg:M:T")
+
+  private def create(builder: TransactionBuilder, id: String, hasKey: Boolean = false) =
+    builder.create(
+      id = id,
+      template = templateId.toString,
+      argument = Value.ValueRecord(None, ImmArray.empty),
+      signatories = Seq(),
+      observers = Seq(),
+      key = if (hasKey) Some(keyValue) else None,
+    )
+
+  private def exercise(
+      builder: TransactionBuilder,
+      id: String,
+      hasKey: Boolean = false,
+      consuming: Boolean = true,
+  ) =
+    builder.exercise(
+      contract = create(builder, id, hasKey),
+      choice = "Choice",
+      argument = Value.ValueRecord(None, ImmArray.empty),
+      actingParties = Set(),
+      consuming = consuming,
+      result = Some(Value.ValueUnit),
+    )
+
+  private def fetch(builder: TransactionBuilder, id: String, byKey: Boolean) =
+    builder.fetch(contract = create(builder, id, hasKey = true), byKey = byKey)
+
+  private def lookup(builder: TransactionBuilder, id: String, found: Boolean) =
+    builder.lookupByKey(contract = create(builder, id, hasKey = true), found = found)
+
+  import Conversions.{contractIdToStateKey, contractKeyToStateKey}
+
+  "submissionOutputs" should {
+    "return a single output for a create without a key" in {
+      val builder = TransactionBuilder()
+      val c = create(builder, "#1")
```

Comment:
Please name this `contract` or something.

Reply:
f35f65028ebcf9d6476e2362cccc187f44b58233